python/semantic_kernel/utils/tracing/decorators.py

```diff
@@ -0,0 +1,233 @@
+import json
+import os
+from collections.abc import Callable
+from typing import Any
+
+from opentelemetry import trace
+from opentelemetry.trace import Span, StatusCode
+
+from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
+from semantic_kernel.contents.chat_history import ChatHistory
+from semantic_kernel.contents.chat_message_content import ITEM_TYPES, ChatMessageContent
+from semantic_kernel.contents.function_call_content import FunctionCallContent
+from semantic_kernel.utils.tracing.const import (
+    COMPLETION_EVENT,
+    COMPLETION_EVENT_COMPLETION,
+    COMPLETION_TOKEN,
+    ERROR_TYPE,
+    FINISH_REASON,
+    MAX_TOKEN,
+    MODEL,
+    OPERATION,
+    PROMPT_EVENT,
+    PROMPT_EVENT_PROMPT,
+    PROMPT_TOKEN,
+    RESPONSE_ID,
+    SYSTEM,
+    TEMPERATURE,
+    TOP_P,
+)
+
+OTEL_ENABLED_ENV_VAR = "SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS"
+OTEL_SENSITIVE_ENABLED_ENV_VAR = "SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS_SENSITIVE"
+
+
+_enable_diagnostics = os.getenv(OTEL_ENABLED_ENV_VAR, "false").lower() in ("true", "1", "t")
+_enable_sensitive_events = os.getenv(OTEL_SENSITIVE_ENABLED_ENV_VAR, "false").lower() in ("true", "1", "t")
+
+# Creates a tracer from the global tracer provider
+tracer = trace.get_tracer(__name__)
+
+
+def are_model_diagnostics_enabled() -> bool:
+    """Check if model diagnostics are enabled.
+
+    Model diagnostics are enabled if either EnableModelDiagnostics or EnableSensitiveEvents is set.
+    """
+    return _enable_diagnostics or _enable_sensitive_events
+
+
+def are_sensitive_events_enabled() -> bool:
+    """Check if sensitive events are enabled.
+
+    Sensitive events are enabled if EnableSensitiveEvents is set.
+    """
+    return _enable_sensitive_events
+
+
+def trace_chat_completion(model_provider: str) -> Callable:
+    """Decorator to trace chat completion activities."""
+
+    def inner_trace_chat_completion(completion_func: Callable) -> Callable:
+        @functools.wraps(completion_func)
+        async def wrapper_decorator(*args: Any, **kwargs: Any) -> list[ChatMessageContent]:
+            chat_history: ChatHistory = kwargs["chat_history"]
+            settings: PromptExecutionSettings = kwargs["settings"]
+
+            if hasattr(settings, "ai_model_id") and settings.ai_model_id:
+                model_name = settings.ai_model_id
+            elif hasattr(args[0], "ai_model_id") and args[0].ai_model_id:
+                model_name = args[0].ai_model_id
+            else:
+                model_name = "unknown"
+
+            span = _start_completion_activity(model_name, model_provider, chat_history, settings)
+
+            try:
+                completions: list[ChatMessageContent] = await completion_func(*args, **kwargs)
+            except Exception as exception:
+                if span:
+                    _set_completion_error(span, exception)
+                    span.end()
+                raise
```

Comment:
let's make reraise explicit: `raise exception`

Reply:
I'm pretty new to Python.  Won't that make us lose our stack info (like this would do in C#)?

Reply:
if you do something like:

```python
...
except Exception as ex:
    # do span related check
    raise <your_exception_type_here>("Some message here", ex) from ex
```
the `from ex` is used to explicitly chain exceptions, which helps to maintain the context and the stack trace of the original exception while raising a new exception.